### PR TITLE
Backend docstring cleanup

### PR DESCRIPTION
**Pull request by smite**
**Description of the Change:**

Starting with Sphinx 1.7, child class methods without explicit docstrings inherit their docstrings from the parent class method they override. Hence child classes should only provide a docstring if the method functionality has been changed.

The idea is to take all useful information from the backend API child class docstrings and move it to the corresponding BaseBackend, BaseFock or BaseGaussian docstring.

I've fixed the FockBackend  docstrings, the two other backends still need the same treatment.
This is not urgent, and can wait until we've redesigned the backend API.

**Benefits:**

Avoid docstring duplication and conflicting information. For each API call there should be just one authoritative docstring.

### CODE REVIEW

File: strawberryfields/backends/base.py [hunk 1]
```diff
@@ -260,8 +253,8 @@ def supports(self, name):
     def begin_circuit(self, num_subsystems, *, cutoff_dim=None, pure=True, **kwargs):
         r"""Instantiate a quantum circuit.
 
-        Instantiates a circuit with num_subsystems modes to track and update a quantum optical state.
-        The state of the circuit is initialized to vacuum.
+        Instantiates a representation of a quantum optical state with num_subsystems modes.
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
        Instantiates a representation of a quantum optical state with ``num_subsystems`` modes.
```

File: strawberryfields/backends/base.py [hunk 2]
```diff
@@ -370,6 +364,8 @@ def prepare_squeezed_state(self, r, phi, mode):
     def prepare_displaced_squeezed_state(self, alpha, r, phi, mode):
         r"""Prepare a displaced squeezed state in the specified mode.
 
+        FIXME this method is not currently used by the engine
+
```
Comment (josh146) [inline strawberryfields/backends/base.py]: That's strange :thinking:

File: strawberryfields/backends/base.py [hunk 3]
```diff
@@ -475,38 +473,40 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
...
 
     def is_vacuum(self, tol=0.0, **kwargs):
-        r"""Test whether the current circuit state is in vacuum (up to tolerance tol).
+        r"""Test whether the current circuit state is vacuum (up given tolerance).
+
+        FIXME The backends do not implement this method consistently, fockbackend uses np.linalg.norm whereas the other two use fidelity. Switch to fidelity for all?
 
         Args:
-            tol (float): numerical tolerance for how close state must be to true vacuum state
+            tol (float): numerical tolerance for how close state must be to true vacuum state (in terms of fidelity?)
```
Comment (josh146) [inline strawberryfields/backends/base.py]: Let's fix this is up before merging this PR, so that it doesn't appear in the documentation

File: strawberryfields/backends/base.py [hunk 4]
```diff
@@ -475,38 +473,40 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
...
 
     def state(self, modes=None, **kwargs):
         r"""Returns the state of the quantum simulation, restricted to the subsystems defined by `modes`.
 
         Args:
-            modes (int or Sequence[int]): specifies the mode(s) to restrict the return state to
-                This argument is optional; the default value ``modes=None`` returns the state containing all modes.
+            modes (int or Sequence[int] or None): Specifies the modes to restrict the return state to.
+                None returns the state containing all the modes.
                 If modes is not ordered, the returned state contains the requested modes in the given order, i.e.,
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                If ``modes`` is not ordered, the returned state contains the requested modes in the given order, i.e.,
```

File: strawberryfields/backends/base.py [hunk 5]
```diff
@@ -475,38 +473,40 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
...
                 requesting the modes=[3,1] results in a two mode state being returned with the first mode being
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                requesting ``modes=[3,1]`` results in a two mode state being returned with the first mode being
```

File: strawberryfields/backends/base.py [hunk 6]
```diff
@@ -642,9 +643,9 @@ def measure_fock(self, modes, select=None, **kwargs):
 
         Args:
             modes (Sequence[int]): which modes to measure
-            select (Sequence[int]): (Optional) desired values of measurement results.
-                Allows user to post-select on specific measurement results instead of randomly sampling.
-
+            select (None or Sequence[int]): If not None: desired values of the measurement results.
+                Enables post-selection on specific measurement results instead of random sampling.
+                len(select) == len(modes).
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                Note that ``len(select) == len(modes)``.
```

File: strawberryfields/backends/base.py [hunk 7]
```diff
@@ -654,8 +655,11 @@ def state(self, modes=None, **kwargs):
         r"""Returns the state of the quantum simulation, restricted to the subsystems defined by `modes`.
 
         Args:
-            modes (int or Sequence[int]): specifies the mode or modes to restrict the return state to.
-                This argument is optional; the default value ``modes=None`` returns the state containing all modes.
+            modes (int or Sequence[int] or None): Specifies the modes to restrict the return state to.
+                None returns the state containing all the modes.
+                If modes is not ordered, the returned state contains the requested modes in the given order, i.e.,
+                requesting the modes=[3,1] results in a two mode state being returned with the first mode being
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                requesting ``modes=[3,1]`` results in a two mode state being returned with the first mode being
```

File: strawberryfields/backends/base.py [hunk 8]
```diff
@@ -654,8 +655,11 @@ def state(self, modes=None, **kwargs):
         r"""Returns the state of the quantum simulation, restricted to the subsystems defined by `modes`.
 
         Args:
-            modes (int or Sequence[int]): specifies the mode or modes to restrict the return state to.
-                This argument is optional; the default value ``modes=None`` returns the state containing all modes.
+            modes (int or Sequence[int] or None): Specifies the modes to restrict the return state to.
+                None returns the state containing all the modes.
+                If modes is not ordered, the returned state contains the requested modes in the given order, i.e.,
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                If ``modes`` is not ordered, the returned state contains the requested modes in the given order, i.e.,
```

File: strawberryfields/backends/base.py [hunk 9]
```diff
@@ -703,41 +707,47 @@ def prepare_gaussian_state(self, r, V, modes):
         Args:
             r (array): vector of means in xp ordering
             V (array): covariance matrix in xp ordering
-            modes (int or Sequence[int]): which mode to prepare the state in
-                If the modes are not sorted, this is take into account when preparing the state.
-                i.e., when a two mode state is prepared in modes=[3,1], then the first
-                mode of state goes into mode 3 and the second mode goes into mode 1 of the simulator.
+            modes (int or Sequence[int]): Which modes to prepare the state in.
+                If the modes are not sorted, this is taken into account when preparing the state.
+                I.e., when a two mode state is prepared in modes=[3,1], the first
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                I.e., when a two mode state is prepared with ``modes=[3,1]``, the first
```

File: strawberryfields/backends/base.py [hunk 10]
```diff
@@ -703,41 +707,47 @@ def prepare_gaussian_state(self, r, V, modes):
...
     def kerr_interaction(self, kappa, mode):
-        # pylint: disable=unused-argument,missing-docstring
         raise NotApplicableError
 
     def cross_kerr_interaction(self, kappa, mode1, mode2):
-        # pylint: disable=unused-argument,missing-docstring
         raise NotApplicableError
 
     def measure_fock(self, modes, select=None):
-        # pylint: disable=unused-argument,missing-docstring
```
Comment (josh146) [inline strawberryfields/backends/base.py]: Nice, pylint didn't used to be able to determine this automatically (must be due to keeping up with the new Sphinx functionality).

File: strawberryfields/backends/base.py [hunk 11]
```diff
@@ -703,41 +707,47 @@ def prepare_gaussian_state(self, r, V, modes):
...
         raise NotApplicableError
+
+    def state(self, modes=None, **kwargs):
```
Comment (josh146) [inline strawberryfields/backends/base.py]: Was this not present previously?

File: strawberryfields/backends/base.py [hunk 12]
```diff
@@ -703,41 +707,47 @@ def prepare_gaussian_state(self, r, V, modes):
...
+        r"""Returns the state of the quantum simulation, restricted to the subsystems defined by `modes`.
+
+        Args:
+            modes (int or Sequence[int] or None): Specifies the modes to restrict the return state to.
+                None returns the state containing all the modes.
+                If modes is not ordered, the returned state contains the requested modes in the given order, i.e.,
```
Comment (josh146) [inline strawberryfields/backends/base.py]: Note: should this instead say 'If modes *is* ordered`? If so, needs to be updated in all other docstrings this sentence appears in.

```suggestion
                If ``modes`` is not ordered, the returned state contains the requested modes in the given order, i.e.,
```

File: strawberryfields/backends/base.py [hunk 13]
```diff
@@ -703,41 +707,47 @@ def prepare_gaussian_state(self, r, V, modes):
...
+                requesting the modes=[3,1] results in a two mode state being returned with the first mode being
```
Comment (josh146) [inline strawberryfields/backends/base.py]:
```suggestion
                requesting ``modes=[3,1]`` results in a two mode state being returned with the first mode being
```

File: strawberryfields/backends/fockbackend/backend.py [hunk 14]
```diff
@@ -83,181 +83,53 @@ def begin_circuit(self, num_subsystems, *, cutoff_dim=None, pure=True, **kwargs)
...
-            r (complex): reflectivity parameter
-            mode1 (int): index of first mode where operation is carried out
-            mode2 (int): index of second mode where operation is carried out
-
-        """
         if isinstance(t, complex):
             raise ValueError("Beamsplitter transmittivity t must be a float.")
         self.circuit.beamsplitter(t, abs(r), phase(r), self._remap_modes(mode1), self._remap_modes(mode2))
 
-    def kerr_interaction(self, kappa, mode):
```
Comment (josh146) [inline strawberryfields/backends/fockbackend/backend.py]: Why is this deleted?

Comment (josh146) [reply strawberryfields/backends/fockbackend/backend.py]: Oh, I see you've moved it to the bottom.

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 15]
```diff
@@ -27,198 +27,112 @@
...
+        self._init_modes = None
+        self.circuit = None
 
     def begin_circuit(self, num_subsystems, *, cutoff_dim=None, pure=None, **kwargs):
-        r"""
-        Create a quantum circuit (initialized in vacuum state) with number of modes
-        equal to num_subsystems.
+        r"""Instantiate a quantum circuit.
+
+        Instantiates a representation of a quantum optical state with num_subsystems modes.
```
Comment (josh146) [inline strawberryfields/backends/gaussianbackend/backend.py]:
```suggestion
        Instantiates a representation of a quantum optical state with ``num_subsystems`` modes.
```

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 16]
```diff
@@ -27,198 +27,112 @@
...
-        Note: this may convert the state representation to mixed.
-
-        Args:
-            alpha (complex): displacement parameter
-            r (float): squeezing amplitude
-            phi (float): squeezing phase
-            mode (int): index of mode where state is prepared
-        """
-
+        # FIXME reset to vacuum missing!
```
Comment (josh146) [inline strawberryfields/backends/gaussianbackend/backend.py]: Let's do this in this PR, or make a new issue for it

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 17]
```diff
@@ -27,198 +27,112 @@
...
-                                Perfect homodyning is obtained when eps \to 0.
+            phi (float): phase angle of the quadrature to measure (x: :math:`\phi=0`, p: :math:`\phi=\pi/2`)
+            mode (int): which mode to measure
+            select (None or float): If not None: desired value of the measurement result.
+                Enables post-selection on specific measurement results instead of random sampling.
+
+        Keyword Args:
+            eps (float): Homodyne amounts to projection onto a quadrature eigenstate.
+                This eigenstate is approximated by a squeezed state whose variance has been
+                squeezed to the amount eps, V_(meas) = eps**2.
```
Comment (josh146) [inline strawberryfields/backends/gaussianbackend/backend.py]:
```suggestion
                squeezed to the amount ``eps``, :math:`V_{meas} = \texttt{eps}^2`.
```

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 18]
```diff
@@ -27,198 +27,112 @@
...
-            float or Tensor: measurement outcomes
+            float: measured value
         """
-        if "eps" in kwargs:
-            eps = kwargs["eps"]
-        else:
-            eps = 0.0002
+        eps = kwargs.get('eps', 0.0002)
 
+        # phi is the rotation of the measurement operator, hence the minus
```
Comment (josh146) [inline strawberryfields/backends/gaussianbackend/backend.py]: thumb

File: strawberryfields/backends/tfbackend/backend.py [hunk 19]
```diff
@@ -56,27 +59,31 @@ def _remap_modes(self, modes):
...
         if was_int:
             remapped_modes = remapped_modes[0]
         return remapped_modes
 
     def begin_circuit(self, num_subsystems, *, cutoff_dim=None, pure=True, **kwargs):
-        r"""Create a quantum circuit (initialized in vacuum state) with the number of modes
-        equal to num_subsystems and a Fock-space cutoff dimension of cutoff_dim.
+        r"""Instantiate a quantum circuit.
+
+        Instantiates a representation of a quantum optical state with num_subsystems modes.
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
        Instantiates a representation of a quantum optical state with ``num_subsystems`` modes.
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 20]
```diff
@@ -56,27 +59,31 @@ def _remap_modes(self, modes):
...
+        The state is initialized to vacuum.
+
+        The modes in the circuit are indexed sequentially using integers, starting from zero.
+        Once an index is assigned to a mode, it can never be re-assigned to another mode.
+        If the mode is deleted its index becomes invalid.
+        An operation acting on an invalid or unassigned mode index raises an IndexError exception.
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
        An operation acting on an invalid or unassigned mode index raises an ``IndexError`` exception.
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 21]
```diff
@@ -362,12 +227,19 @@ def state(self, modes=None, **kwargs):
         r"""Returns the state of the quantum simulation, restricted to the subsystems defined by `modes`.
 
         Args:
-            modes (int or Sequence[int]): specifies the mode or modes to restrict the return state to.
-                                          This argument is optional; the default value ``modes=None`` returns the state containing all modes.
-            **kwargs: optional keyword args (`session`: a Tensorflow session; `feed_dict`: a Python dictionary feeding the desired numerical values for Tensors)
-                      which will be used by the Tensorflow simulator for numerically evaluating the measurement results.
+            modes (int or Sequence[int] or None): Specifies the modes to restrict the return state to.
+                None returns the state containing all the modes.
+                If modes is not ordered, the returned state contains the requested modes in the given order, i.e.,
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]: If modes **is** ordered?

```suggestion
                If ``modes`` is ordered, the returned state contains the requested modes in the given order, i.e.,
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 22]
```diff
@@ -362,12 +227,19 @@ def state(self, modes=None, **kwargs):
...
+                requesting the modes=[3,1] results in a two mode state being returned with the first mode being
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
                requesting ``modes=[3,1]`` results in a two mode state being returned with the first mode being
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 23]
```diff
@@ -421,14 +293,20 @@ def state(self, modes=None, **kwargs):
...
 
         Args:
-            modes (Sequence[int]): indices of mode where operation is carried out
-            select (Sequence[int]): (Optional) desired values of measurement results. Allows user to post-select on specific measurement results instead of randomly sampling.
-            **kwargs: optional keyword args (`session`: a Tensorflow session; `feed_dict`: a Python dictionary feeding the desired numerical values for Tensors)
-                                which will be used by the Tensorflow simulator for numerically evaluating the measurement results.
+            modes (Sequence[int]): which modes to measure
+            select (None or Sequence[int]): If not None: desired values of the measurement results.
+                Enables post-selection on specific measurement results instead of random sampling.
+                len(select) == len(modes).
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
                Note that ``len(select) == len(modes)``.
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 24]
```diff
@@ -421,14 +293,20 @@ def state(self, modes=None, **kwargs):
...
+
+        Keyword Args:
+            session (tf.Session): TensorFlow session
+            feed_dict (Dict): Dictionary containing the desired numerical values for Tensors
+                for numerically evaluating the measurement results. Used with `session`.
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
                for numerically evaluating the measurement results. Used with ``session``.
```

File: strawberryfields/backends/tfbackend/backend.py [hunk 25]
```diff
@@ -446,11 +324,14 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
...
-            **kwargs: optional keyword args (`session`: a Tensorflow session; `feed_dict`: a Python dictionary feeding the desired numerical values for Tensors)
-                                which will be used by the Tensorflow simulator for numerically evaluating the measurement results.
-                                In addition, kwargs can be used to (optionally) pass user-specified numerical parameters `max` and `num_bins`.
-                                These are used numerically to build the probability distribution function (pdf) for the homodyne measurement.
-                                Specifically, the pdf is discretized onto the 1D grid [-max,max], with num_bins equally spaced bins.
+
+        Keyword Args:
+            session (tf.Session): TensorFlow session
+            feed_dict (Dict): Dictionary containing the desired numerical values for Tensors
+                for numerically evaluating the measurement results. Used with `session`.
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/backend.py]:
```suggestion
                for numerically evaluating the measurement results. Used with ``session``.
```

File: strawberryfields/backends/base.py (hunk 11 above)
Comment (smite) [reply strawberryfields/backends/base.py]: No, but we only need it here to be able to override the docstring, to show that `BaseGaussian.state()` returns a GaussianState.

File: strawberryfields/backends/base.py [hunk 26]
```diff
@@ -475,38 +473,40 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
...
 
         Returns:
-            float: measured value
+            float or array[float]: measured value
         """
         raise NotImplementedError
 
     def is_vacuum(self, tol=0.0, **kwargs):
-        r"""Test whether the current circuit state is in vacuum (up to tolerance tol).
+        r"""Test whether the current circuit state is vacuum (up given tolerance).
```
Comment (co9olguy) [inline strawberryfields/backends/base.py]:
```suggestion
        r"""Test whether the current circuit state is vacuum (up to given tolerance).
```

File: strawberryfields/backends/base.py [hunk 27]
```diff
@@ -581,13 +582,13 @@ def prepare_dm_state(self, state, modes):
         raise NotImplementedError
 
 
-    def cubic_phase(self, gamma_prime, mode):
```
Comment (co9olguy) [inline strawberryfields/backends/base.py]: weird 🤔, wonder where this came from?

File: strawberryfields/backends/fockbackend/backend.py [hunk 28]
```diff
@@ -50,8 +50,8 @@ def _remap_modes(self, modes):
         submap = [map_[m] for m in modes]
```
Comment (co9olguy) [inline strawberryfields/backends/fockbackend/backend.py]: (Throughout): Before merging, let's confirm that these still import correct docstrings in compiled docs

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 29]
```diff
@@ -27,198 +27,112 @@
...
+        self.circuit = None
 
     def begin_circuit(self, num_subsystems, *, cutoff_dim=None, pure=None, **kwargs):
-        r"""
-        Create a quantum circuit (initialized in vacuum state) with number of modes
-        equal to num_subsystems.
+        r"""Instantiate a quantum circuit.
+
+        Instantiates a representation of a quantum optical state with num_subsystems modes.
+        The state is initialized to vacuum.
```
Comment (co9olguy) [inline strawberryfields/backends/gaussianbackend/backend.py]:
```suggestion
        The state is initialiazed to vacuum.
```

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 30]
```diff
@@ -27,198 +27,112 @@
...
+
+        The modes in the circuit are indexed sequentially using integers, starting from zero.
+        Once an index is assigned to a mode, it can never be re-assigned to another mode.
+        If the mode is deleted its index becomes invalid.
+        An operation acting on an invalid or unassigned mode index raises an IndexError exception.
 
         Args:
-            num_subsystems (int): number of modes the circuit should begin with
+            num_subsystems (int): number of modes in the circuit
+            cutoff_dim (int): unused in this backend
```
Comment (co9olguy) [inline strawberryfields/backends/gaussianbackend/backend.py]: Do these unused args show up somewhere in the compiled docs, necessitating an extra explainer here?

File: strawberryfields/backends/tfbackend/backend.py [hunk 31]
```diff
@@ -45,6 +44,10 @@ def __init__(self, graph=None):
             self._graph = tf.get_default_graph()
         else:
             self._graph = graph
+        # TODO docstrings
```
Comment (co9olguy) [inline strawberryfields/backends/tfbackend/backend.py]: Is this not part of current PR?

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 29 above)
Comment (josh146) [reply strawberryfields/backends/gaussianbackend/backend.py]: I don't think this is right? 'initialized' should be correct

**Comment by josh146** [comment]
Also: the `FIXME`'s will show up in the online docs --- they should either be fixed here, or moved to a comment below the docstring

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 29 above)
Comment (co9olguy) [reply strawberryfields/backends/gaussianbackend/backend.py]: lol you're right! Please ignore

File: strawberryfields/backends/base.py (hunk 2 above)
Comment (smite) [reply strawberryfields/backends/base.py]: The `DisplacedSqueezed` Operation uses `backend.prepare_squeezed_state` and then `backend.displacement`, not this method directly.

File: strawberryfields/backends/base.py (hunk 12 above)
Comment (smite) [reply strawberryfields/backends/base.py]: Actually it should say that the returned state contains the modes in the same order they are listed in `modes`. I'll change it.

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 30 above)
Comment (smite) [reply strawberryfields/backends/gaussianbackend/backend.py]: The Fock backends use them, and they appear in the method call signature so some documentation is needed. Pylint frowns upon changing the call signature when overriding a parent class method so they can't just be removed. One solution would be to move them explicitly into kwargs (that's how the Engine already treats them).

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 16 above)
Comment (smite) [reply strawberryfields/backends/gaussianbackend/backend.py]: I'll just add the reset there, that should fix it.

File: strawberryfields/backends/tfbackend/backend.py (hunk 31 above)
Comment (smite) [reply strawberryfields/backends/tfbackend/backend.py]: Can be, but I'm not 100% familiar with the TF backend so if I write the docstrings they may be not perfectly accurate. Do you want to add them?

File: strawberryfields/backends/base.py (hunk 3 above)
Comment (smite) [reply strawberryfields/backends/base.py]: Do you think fidelity would be a good measure here? @co9olguy

File: strawberryfields/backends/base.py (hunk 2 above)
Comment (co9olguy) [reply strawberryfields/backends/base.py]: I think it should do what every other Operation does (at least with a local execution), i.e., try to call directly the appropriate backend method, then decompose as needed upon failure

File: strawberryfields/backends/base.py (hunk 3 above)
Comment (co9olguy) [reply strawberryfields/backends/base.py]: yes, especially if 2/3 backends already use it

File: strawberryfields/backends/tfbackend/backend.py (hunk 31 above)
Comment (co9olguy) [reply strawberryfields/backends/tfbackend/backend.py]: why don't you have a go at writing them, then I can edit and update as needed?

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 30 above)
Comment (josh146) [reply strawberryfields/backends/gaussianbackend/backend.py]: I'm okay with the solution used above, as it satisfies the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)

File: strawberryfields/backends/base.py (hunk 2 above)
Comment (josh146) [reply strawberryfields/backends/base.py]: :+1:

So, something like the following?

```python
    def _apply(self, reg, backend, **kwargs):
        p = _unwrap(self.p)
        # prepare the displaced squeezed state directly
        backend.prepare_displaced_squeezed_state(p[0], p[1], p[2], *reg)

    def _decompose(self, reg):
        # decompoe into a squeezed state and a displacement
        return [
            Command(Squeezed(self.p[1], self.p[2]), reg),
            Command(Dgate(self.p[0]), reg)
        ]
```

File: strawberryfields/backends/base.py [hunk 32]
```diff
@@ -453,7 +453,7 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
         r"""Measure a :ref:`phase space quadrature <homodyne>` of the given mode.
 
         For the measured mode, samples the probability distribution
-        :math:`f(q) = \bra{q_\phi} R^\dagger(\phi) \rho R(\phi) \ket{q_\phi}`
```
Comment (co9olguy) [inline strawberryfields/backends/base.py]: good find

File: strawberryfields/backends/gaussianbackend/backend.py [hunk 33]
```diff
@@ -80,6 +80,11 @@ def prepare_squeezed_state(self, r, phi, mode):
         self.circuit.loss(0.0, mode)
         self.circuit.squeeze(r, phi, mode)
 
+    def prepare_displaced_squeezed_state(self, alpha, r, phi, mode):
+        self.circuit.loss(0.0, mode)
```
Comment (co9olguy) [inline strawberryfields/backends/gaussianbackend/backend.py]: why is `loss` included here?

Comment (josh146) [reply strawberryfields/backends/gaussianbackend/backend.py]: Yep, that is how @nquesada does resets in the Gaussian backend

File: strawberryfields/backends/tfbackend/backend.py (hunk 31 above)
Comment (smite) [reply strawberryfields/backends/tfbackend/backend.py]: Sure.

File: strawberryfields/backends/base.py (hunk 3 above)
Comment (smite) [reply strawberryfields/backends/base.py]: Fidelity or distance in terms of the https://en.wikipedia.org/wiki/Bures_metric ?

Comment (smite) [reply strawberryfields/backends/base.py]: The difference seems to be a square root and a scaling of tol by 2

Comment (co9olguy) [reply strawberryfields/backends/base.py]: Just default to whatever the current "majority vote" is amongst the backends that agree

File: strawberryfields/backends/gaussianbackend/backend.py (hunk 30 above)
Comment (smite) [reply strawberryfields/backends/gaussianbackend/backend.py]: I changed them into kwargs, now the docstrings look nice.

File: strawberryfields/backends/base.py (hunk 3 above)
Comment (smite) [reply strawberryfields/backends/base.py]: Fidelity it is then.

File: strawberryfields/backends/base.py (hunk 2 above)
Comment (smite) [reply strawberryfields/backends/base.py]: Fixed as suggested.

**Comment by co9olguy** [comment]
@smite Let us know when the final issues/changes are implemented and this is ready for final review

**Comment by smite** [comment]
@co9olguy It should be ready now, once the tests finish.
